year20/src/day05.rs

```diff
@@ -1,9 +1,49 @@
-// Potential improvements:
-//
+fn str_to_decimal(s: &str, one_char: char, zero_char: char) -> u64 {
+    // Map the string slice to a binary number as a String.
+    let binary_string = s
+        .chars()
+        .map(|c| match c {
+            one if one == one_char => '1',
+            zero if zero == zero_char => '0',
+            _ => panic!(
+                "Character in seat address was not {} or {}",
+                one_char, zero_char
+            ),
+        })
+        .collect::<String>();
+    // Then convert this binary string to an integer.
+    u64::from_str_radix(&binary_string, 2).unwrap()
+}
+
+pub fn day05(input_lines: &[Vec<String>]) -> (String, String) {
+    // The seat address can be converted to a row and column by splitting it into it's two parts, converting
+    // those parts into a binary string  and then converting those binary numbers into integers. Then we can
+    // use the row and column to find the seat ID with row * 8 + col.
+    // E.g., "FBFBBFFRLR" -> "FBFBBFF", "RLR" -> "0101100", "101" -> 44, 5 -> 44 * 8 + 5 = 357
+    let mut seat_ids: Vec<u64> = input_lines[0]
+        .iter()
+        .map(|line| {
+            let (row, col) = line.split_at(7);
+            str_to_decimal(row, 'B', 'F') * 8 + str_to_decimal(col, 'R', 'L')
+        })
+        // I could have called .max().unwrap() here for part one rather than sorting the Vec and taking the last element.
+        // However, for part two, I needed to have a sorted Vec anyway.
+        .collect();
+
+    seat_ids.sort();
+
+    let answer1: &u64 = seat_ids.last().unwrap();
+
+    let mut answer2: u64 = 0;
+    for id_window in seat_ids.windows(2) {
```

Comment:
I think `windows()` here is slightly overkill, but my arguments for why are only really concerns for much larger data-sets than we have here!

Reasons why it's slightly inefficient:
- We're doubling the memory used by this Vec for the duration of this loop (well, trebling since we still have the Vec independently).  If we just walked an iterator of the vec, keeping the last value looked at in a separate variable, we avoid this.
- We're creating the windows over the entire list, regardless of how early we find the thing we're looking for.  Note that with the check you're using we're walking the whole thing regardless, so that's unavoidable, but with a change to that we could also avoid even doing the extra work on putting all the list elements into windows.